webhooks: Allow event registration using webhook_view decorator.

PIG208 opened the pull request:
Reopened #18525

timabbott commented on zerver/webhooks/airbrake/tests.py:
```diff
@@ -5,6 +5,9 @@ class AirbrakeHookTests(WebhookTestCase):
     WEBHOOK_DIR_NAME = "airbrake"
+    # We ensure that even we don't register any events in the view function
+    # this WebhookTestCase will not raise an exception when send_webhook_message is called
+    VIEW_FUNCTION_NAME = "api_airbrake_webhook"
```
Hmm, I don't see an exception if I delete this line and run tests.

PIG208 replied:
This line is meant for test coverage in `zerver/lib/test_classes.py` at line 1412
```
if all_event_types is None:
        return
```

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -1384,6 +1394,55 @@ def setUp(self) -> None:
+
+It is likely that you used the incoming webhook event filtering system without adding this
+event to all_event_types in the webhook_view function decorator, or you incorrectly
+parsed the event type when calling check_send_webhook_message.
```
I think this is confusing, in that the developer may not have used the incoming webhook event filtering system, if they're adding new events to an existing webhook.

Proposed new language:
```
Error: This test triggered a message using the event "{complete_event_type}", which was not properly
registered via the @webhook_view(..., event_types=[...]). These registrations are important for Zulip
self-documenting the supported event types for this integration.

You can fix this by adding "{complete_event_type}" to ALL_EVENT_TYPES for this webhook. 
See {documentation_url} for details.
```

(We can leave off the last line for now, and add it when we add documentation for this infrastructure).

And then once we've added the docs and event_types declarations for most webhooks, we can have a separate exception for "webhook not on an exclude list of webhooks that are known to not declare event types", to prevent new webhooks from being introduced without this feature.

PIG208 replied:
OK, this sounds reasonable.

timabbott commented:
I posted a couple comments, but otherwise this looks good.  Once we've merged this, I guess we can proceed in parallel with adding these registrations to webhooks (I expect `test-backend --nonfatal-errors zerver/webhooks/foo/` to be a helpful pattern) and adding something on the /integrations docs to document the new feature.

I'd prioritize doing GitHub next on the registration front, since we'd use that on chat.zulip.org (and thus get to test the system early).

PIG208 commented:
The PR is updated accordingly.

I have removed the `VIEW_FUNCTION_NAME` line in airbrakes' test case and instead mark the line as `nocoverage`. I have also updated the error message without the last line.

timabbott commented:
Great, merged, after a pass on the comments in the test class logic.

I think probably if we're clever about how to do the lookups, we could figure out how to remove the need to specify `VIEW_FUNCTION_NAME`, but maybe it's not worth the effort since it's rare that we need it (and probably in most cases we can just rename the view function).

We can now focus on the parallel projects of adding these tags for all of our existing webhooks (@adambirds maybe that's be a good project to tag-team on) and adding logic to support displaying this feature in the documentation.

PIG208 commented:
Errors for not listed will not be raised at all during tests if `VIEW_FUNCTION_NAME` is not explicitly specified in the test case, which is currently an implicit logic.

One thing to note is that we don't fallback to the default "api_{WEBHOOK_DIR_NAME}_webhook" when doing lookups for the view function. It will cause `ImportError` for webhooks that do not follow the naming convention, and we will need to either suppress that error or specify `VIEW_FUNCTION_NAME` for those webhooks. I prefer the latter approach, but I think we can work on that after migrating most of the webhooks.
